Orange/widgets/visualize/utils/__init__.py

```diff
@@ -351,23 +327,53 @@ def toggle(self):
             self.button.setText("Pause")
             self.progressBarInit()
             self.progress_timer.start()
+            self.before_running()
             self._thread.start()
         else:
             self.button.setText("Continue")
             self._thread.quit()
 
+    def before_running(self):
+        """Code that is run before running vizrank in its own thread"""
+        pass
+
+    def stopped(self):
+        """Code that is run after stopping the vizrank thread"""
+        pass
+
 
 class Worker(QObject):
     done = Signal()
     stopped = Signal()
 
-    def __init__(self, function):
+    def __init__(self, obj):
         super().__init__()
-        self.function = function
+        self.obj = obj
 
     @Slot()
     def do_work(self):
-        self.function()
+        for state in self.obj.iterate_states(self.obj.saved_state):
+            self.obj.saved_state = state
+            if not self.obj.keep_running:
+                self.stopped.emit()
+                return
+            try:
+                score = self.obj.compute_score(state)
+                if score is not None:
+                    pos = bisect_left(self.obj.scores, score)
+                    row_items = self.obj.row_for_state(score, state)
+                    if self.obj._has_bars:
+                        bar = self.obj.bar_length(score)
+                        if bar is not None:
+                            row_items[0].setData(bar,
+                                                 gui.TableBarItem.BarRole)
+                    self.obj.rank_model.insertRow(pos, row_items)
+                    self.obj.scores.insert(pos, score)
+            except IOError:
```

Comment:
Whoops. I did this for debugging so I could actually see the exception and put here something that could not happen at all. :) I'll change it into an undefined exception as it was! Thanks!